thorlcr/activities/indexread/thindexreadslave.cpp

```diff
@@ -663,45 +667,39 @@ class CIndexReadSlaveActivity : public CIndexReadSlaveBase
         ActivityTimer s(totalCycles, timeActivities);
         choosenLimit = (rowcount_t)helper->getChooseNLimit();
 
-        PARENT::start();
-
         needTransform = helper->needTransform();
-        helperKeyedLimit = (rowcount_t)helper->getKeyedLimit();
+        keyedLimit = (rowcount_t)helper->getKeyedLimit();
         rowLimit = (rowcount_t)helper->getRowLimit(); // MORE - if no filtering going on could keyspan to get count
         if (0 != (TIRlimitskips & helper->getFlags()))
             rowLimit = RCMAX;
         if (!helper->canMatchAny())
-            helperKeyedLimit = keyedLimit = RCMAX; // disable
+            keyedLimit = RCMAX; // disable
         else if (keyedLimit != RCMAX)
         {
             if (TIRkeyedlimitskips & helper->getFlags())
                 keyedLimitSkips = true;
         }
         first = true;
-        keyedLimit = helperKeyedLimit;
         keyedLimitCount = RCMAX;
         keyedProcessed = 0;
         if (steppedExtra)
             steppingMeta.setExtra(steppedExtra);
 
-        if (choosenLimit && !helper->transformMayFilter())
+        // NB: setup remoteLimit before base start() call which if parts remote, will use remoteLimit
+        if (choosenLimit)
         {
-            remoteLimit = keyedLimit;
-            if (keyedLimit && (keyedLimit < remoteLimit))
+            remoteLimit = choosenLimit;
+            if (!helper->transformMayFilter() && (RCMAX != keyedLimit) && (keyedLimit+1 < remoteLimit))
                 remoteLimit = keyedLimit+1; // 1 more to ensure triggered when received back.
         }
-        if (currentInput)
-        {
+
         if ((keyedLimit != RCMAX && (keyedLimitSkips || (helper->getFlags() & TIRcountkeyedlimit) != 0)))
-                keyedLimitCount = getCount(keyedLimit, true);
-        }
-        else if ((keyedLimit != RCMAX && (keyedLimitSkips || (helper->getFlags() & TIRcountkeyedlimit) != 0)))
-        {
-            eoi = false;
-            keyedLimitCount = 0;
-        }
-        else
-            eoi = true;
+            keyedLimitCount = getCount(keyedLimit, true);
+
+        PARENT::start();
+
+        if (eoi && RCMAX != keyedLimitCount && !keyedLimitSkips && (container.queryLocalOrGrouped() || firstNode()))
```

Comment:
Possibly add a comment
//keyed limits are reported on the first node - and might occur even if that node contains no rows.

I think queryLocalOrGroup() is not necessary since keyedLimitCount will always be 0, but possibly better for consistency.